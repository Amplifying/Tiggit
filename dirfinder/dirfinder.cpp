#include "dirfinder.hpp"

#include <boost/filesystem.hpp>
#include <stdlib.h>
#include <fstream>

using namespace DirFinder;
using namespace boost::filesystem;

#ifdef _WIN32

#include <windows.h>
#include <shlobj.h>
#include <tchar.h>

TCHAR pathbuf[MAX_PATH];

path getPathCSIDL(int csidl)
{
  SHGetFolderPath(NULL, csidl, NULL, 0, pathbuf);
  return path(pathbuf);
}

std::string getDataDir(int csidl, const std::string &vname, const std::string &aname)
{
  path p = getPathCSIDL(csidl);
  if(vname != "") p /= vname;
  return (p / aname).string();
}

bool Finder::getStandardPath(std::string &dir)
{
  dir = getDataDir(CSIDL_APPDATA, vname, aname);
  return isWritable(dir);
}

bool openRegKey(const std::string &vname, const std::string &aname, HKEY &hkey)
{
  LPCTSTR sk = ("Software\\" + vname + "\\" + aname).c_str();;

  LONG openRes = RegCreateKeyEx(HKEY_CURRENT_USER,
                                sk,
                                0,NULL,0,0,NULL,
                                &hkey,
                                NULL);

  return openRes == ERROR_SUCCESS;
}


static bool os_setStoredPath(const std::string &dir,
                             const std::string &vname,
                             const std::string &aname,
                             const std::string &dname)
{
  HKEY hkey;
  if(!openRegKey(vname,aname,hkey))
    return false;

  DWORD dwSize = dir.size();
  if(RegSetValueEx(hkey, dname.c_str(), 0, REG_SZ,
                   dir.c_str(), dwSize) != ERROR_SUCCESS)
    return false;
 
  RegCloseKey(hkey);
  return true;
}

static std::string os_getStoredPath(const std::string &vname,
                                    const std::string &aname,
                                    const std::string &dname)
{
  HKEY hkey;
  if(!openRegKey(vname,aname,hkey))
    return false;

  DWORD dwType = REG_SZ;
  DWORD dwSize = MAX_PATH;
  if(RegQueryValueEx(hkey, dname.c_str(), 0, &dwType, 
                     pathbuf, &dwSize) != ERROR_SUCCESS)
    return "";

  RegCloseKey(hkey);

  return string(pathbuf, dwSize);
}

#else
#ifdef __unix

static std::string getHome(const std::string &aname)
{
  return std::string(getenv("HOME")) + "/." + aname + "/";
}

static std::string getPathFile(const std::string &aname,
                               const std::string &dname)
{
  return getHome(aname) + dname + ".conf";
}

bool Finder::getStandardPath(std::string &dir)
{
  dir = getHome(aname);
  return isWritable(dir);
}

static std::string os_getStoredPath(const std::string &vname,
                                    const std::string &aname,
                                    const std::string &dname)
{
  std::string file = getPathFile(aname,dname);
  if(exists(file))
    {
      std::ifstream inp(file.c_str());
      if(inp)
        {
          std::string res;
          inp >> res;
          if(inp)
            return res;
        }
    }
  return "";
}

static bool os_setStoredPath(const std::string &dir,
                             const std::string &vname,
                             const std::string &aname,
                             const std::string &dname)
{
  std::ofstream out(getPathFile(aname,dname).c_str());
  if(out)
    {
      out << dir;
      out.flush();
      if(out)
        return true;
    }
  return false;
}

#else
#error "This platform is not supported yet."
#endif
#endif

bool Finder::getStoredPath(std::string &dir)
{
  dir = os_getStoredPath(vname,aname,dname);
  return isWritable(dir);
}

bool Finder::setStoredPath(const std::string &dir)
{
  return isWritable(dir) && os_setStoredPath(dir,vname,aname,dname);
}

bool Finder::isWritable(const std::string &dir)
{
  if(dir == "")
    return false;

  try
    {
      // Create directories
      create_directories(dir);

      // Do write check
      path p = dir;
      p /= "xyz.tmp";
      const char *file = p.string().c_str();
      {
        std::ofstream out(file);
        if(!out) return false;
        out << "abc";
        if(!out) return false;
      }
      {
        std::ifstream inp(file);
        if(!inp) return false;
        std::string tmp;
        inp >> tmp;
        if(!inp || tmp != "abc")
          return false;
      }

      remove(p);
    }
  // All errors means the dir is not writable
  catch(...){ return false; }
  return true;
}
