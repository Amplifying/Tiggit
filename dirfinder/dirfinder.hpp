#ifndef __DIRFINDER_HPP_
#define __DIRFINDER_HPP_

#include <string>

/*
  This class find an appropriate writable data path for the
  application.

  It can propose platform-dependent standard paths, or accept a
  user-provided value. It will remember settings between sessions.

  We only accept directories that are writable, and we test dirs for
  writability.

  You can have multiple data directories and store independent
  settings for each, if you provide different names to the
  constructor. The default name is 'default'.

  Usage:

  A good practice is to call getStoredPath() first, and ALWAYS prompt
  the user if it returns false. The path returned by getStandardPath()
  may then be presented to the user as a default value.

  Whatever the user supplies, return it to DirFinder through
  setStoredPath().
 */

namespace DirFinder
{
  struct Finder
  {
    /* The vendor_name, app_name and dirname should be filesystem and
       windows registry compatible names. Preferably short, lower-case
       and without spaces.
     */
    Finder(const std::string &vendor_name,
           const std::string &app_name,
           const std::string &dirname = "default")
      : vname(vendor_name), aname(app_name), dname("path-" + dirname) {}

    /* Get stored path. Returns true if found and writable, false
       otherwise.

       If a path is set but the return value is false, it means the
       stored path was invalid or not writable.
     */
    bool getStoredPath(std::string &path);

    /* Get OS-dependent standard path for applications. Will only
       return true if an appropriate writable path was found.

       This does NOT change the stored path, but it may create the
       given path if it does not exist. Use setStoredPath() to use
       this path.
     */
    bool getStandardPath(std::string &path);

    /* Set stored path. Will return true if the path was accepted,
       ie. it was writable, and we were able to store the information
       in a retrievable way.

       If the path was accepted, future requests to getStoredPath()
       will return this path.
     */
    bool setStoredPath(const std::string &path);

    /* Tests if a given directory is valid and writable as a data
       path. Will create directories if they don't exist.
     */
    static bool isWritable(const std::string &path);

  private:
    std::string vname,aname,dname;
  };
}

#endif
