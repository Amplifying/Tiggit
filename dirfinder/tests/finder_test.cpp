#include "dirfinder.hpp"

#include <iostream>
using namespace std;

DirFinder::Finder fnd("tiggit.net", "tiggit", "finder-test");

void setTest(const string &dir)
{
  cout << "Attempting to set standard path to '" << dir << "': ";
  if(fnd.setStoredPath(dir))
    cout << "SUCCESS\n";
  else
    cout << "FAILURE\n";
}

int main()
{
  string out;

  if(fnd.getStoredPath(out))
    cout << "Found stored path: ";
  else
    cout << "No stored path found, out = ";
  cout << out << endl;

  if(fnd.getStandardPath(out))
    cout << "Standard path acceptible: ";
  else
    cout << "Standard path NOT acceptible: ";
  cout << out << endl;

  setTest("/blah");
  setTest(out);
  setTest("/");

  if(fnd.getStoredPath(out))  
    cout << "Current stored path: " << out << endl;
  else
    cout << "FAILED getting stored path (out=" << out << ")\n";

  return 0;
}
