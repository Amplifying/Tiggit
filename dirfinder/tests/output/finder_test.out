Found stored path: /home/mortennk/.tiggit/
Standard path acceptible: /home/mortennk/.tiggit/
Attempting to set standard path to '/blah': FAILURE
Attempting to set standard path to '/home/mortennk/.tiggit/': SUCCESS
Attempting to set standard path to '/': FAILURE
Current stored path: /home/mortennk/.tiggit/
